dbms/src/Storages/DeltaMerge/Index/VectorIndexHNSW/Index.cpp

```diff
@@ -274,16 +309,48 @@ std::vector<VectorIndexBuilder::Key> VectorIndexHNSWViewer::search(
 
     PerfContext::vector_search.visited_nodes += visited_nodes;
     PerfContext::vector_search.discarded_nodes += discarded_nodes;
+    return result;
+}
+
+std::vector<VectorIndexViewer::Key> VectorIndexHNSWViewer::search(
+    const ANNQueryInfoPtr & query_info,
+    const RowFilter & valid_rows) const
+{
+    auto result = searchImpl(query_info, valid_rows);
+
+    std::vector<Key> keys(result.size());
+    result.dump_to(keys.data());
 
     // For some reason usearch does not always do the predicate for all search results.
     // So we need to filter again.
     keys.erase(
         std::remove_if(keys.begin(), keys.end(), [&valid_rows](Key key) { return !valid_rows[key]; }),
         keys.end());
-
     return keys;
 }
 
+std::vector<VectorIndexViewer::SearchResult> VectorIndexHNSWViewer::searchWithDistance(
+    const ANNQueryInfoPtr & query_info,
+    const RowFilter & valid_rows) const
+{
+    auto result = searchImpl(query_info, valid_rows);
+
+    std::vector<Key> keys(result.size());
+    std::vector<Distance> distances(result.size());
+    result.dump_to(keys.data(), distances.data());
```

Comment:
Actually you can access result[i] directly without dumping

Reply:
To keep it simple and save memory.

Reply:
Oh, I mean you can have VectorIndexViewer::SearchResult without double copying, like this maybe:

```c++
const size_t result_size = result.size();
std::vector<SearchResult> search_results;
search_results.reserve(result_size);
for (size_t i = 0; i < result_size; ++i) {
    const auto rowid = result[i].member.key;
    if (valid_rows[rowid])
        search_results.emplace_back({ rowid, result[i].distance }); 
}
```